File: QuantifiedStrides/models/training.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Training load / history
# ---------------------------------------------------------------------------

class TrainingHistoryPointSchema(BaseModel):
    date: date
    load: float
    ctl: float
    atl: float
    tsb: float


class HRVHistoryPointSchema(BaseModel):
    date: date
    hrv: float
    baseline: float
    rhr: int | None
    sleep_score: float | None


class WeeklyVolumeSchema(BaseModel):
    week_start: date
    training_days: int
    total_sets: int


# ---------------------------------------------------------------------------
# Workouts
# ---------------------------------------------------------------------------

class WorkoutListItemSchema(BaseModel):
    workout_id: int
    workout_date: date
    sport: str
    workout_type: str | None
    start_time: datetime | None
    end_time: datetime | None
    duration_s: float | None                  # derived from end_time - start_time
    distance_m: float | None
    avg_hr: int | None
    max_hr: int | None
    calories: int | None
    tss: float | None                         # training_stress_score


class WorkoutMetricPointSchema(BaseModel):
    metric_timestamp: datetime | None
    heart_rate: int | None
    pace: float | None                        # min/km
    cadence: float | None
    vertical_oscillation: float | None        # mm
    vertical_ratio: float | None              # %
    stance_time: float | None                 # ms
    power: float | None                       # watts
    latitude: float | None
    longitude: float | None
    altitude: float | None                    # metres
    distance: float | None                    # cumulative metres
    gradient_pct: float | None
    stride_length: float | None               # cm
    grade_adjusted_pace: float | None         # min/km
    body_battery: float | None                # 0–100
    vertical_speed: float | None              # m/s
    speed_ms: float | None                    # m/s
    grade_adjusted_speed_ms: float | None     # m/s
    performance_condition: int | None         # -20 to +20
    respiration_rate: float | None            # breaths/min


class WorkoutDetailSchema(BaseModel):
    workout_id: int
    workout_date: date
    sport: str
    workout_type: str | None
    start_time: datetime | None
    end_time: datetime | None
    duration_s: float | None
    distance_m: float | None
    avg_hr: int | None
    max_hr: int | None
    calories: int | None
    vo2max: float | None                      # vo2max_estimate
    lactate_threshold: int | None             # lactate_threshold_bpm
    z1: int | None                            # time_in_hr_zone_1 (seconds)
    z2: int | None
    z3: int | None
    z4: int | None
    z5: int | None
    elev_gain: float | None                   # elevation_gain
    elev_loss: float | None                   # elevation_loss
    aerobic_te: float | None                  # aerobic_training_effect
    anaerobic_te: float | None                # anaerobic_training_effect
    tss: float | None                         # training_stress_score
    norm_power: float | None                  # normalized_power
    avg_power: float | None
    max_power: float | None
    avg_cadence: float | None                 # avg_running_cadence
    max_cadence: float | None                 # max_running_cadence
    avg_gct: float | None                     # avg_stance_time
    avg_vo: float | None                      # avg_vertical_oscillation
    avg_stride: float | None                  # avg_stride_length
    avg_vr: float | None                      # avg_vertical_ratio
    total_steps: int | None
    location: str | None
    lat: float | None                         # start_latitude
    lon: float | None                         # start_longitude
    metrics: list[WorkoutMetricPointSchema]

File: mobile/mobile/src/api/endpoints/training.ts
import { apiClient } from '../client'

export interface TrainingHistoryPoint {
  date: string
  load: number
  ctl: number
  atl: number
  tsb: number
}

export interface WorkoutListItem {
  workout_id: number
  workout_date: string
  sport: string
  workout_type: string | null
  start_time: string | null
  end_time: string | null
  duration_s: number | null
  distance_m: number | null
  avg_hr: number | null
  max_hr: number | null
  calories: number | null
  tss: number | null
}

export async function fetchTrainingHistory(days = 42): Promise<TrainingHistoryPoint[]> {
  const res = await apiClient.get('/training/history', { params: { days } })
  return res.data
}

export async function fetchRecentWorkouts(days = 14): Promise<WorkoutListItem[]> {
  const res = await apiClient.get('/training/workouts', { params: { days } })
  return res.data
}

export interface HRVHistoryPoint {
  date: string
  hrv: number
  baseline: number
  rhr: number | null
  sleep_score: number | null
}

export interface WeeklyVolumePoint {
  week_start: string
  training_days: number
  total_sets: number
}

export async function fetchHRVHistory(days = 30): Promise<HRVHistoryPoint[]> {
  const res = await apiClient.get('/training/hrv-history', { params: { days } })
  return res.data
}

export async function fetchWeeklyVolume(weeks = 12): Promise<WeeklyVolumePoint[]> {
  const res = await apiClient.get('/training/weekly-volume', { params: { weeks } })
  return res.data
}
